devlib/target_runner.py

```diff
@@ -0,0 +1,267 @@
+        except subprocess.TimeoutExpired:
+            self.logger.info('Killing target runner...')
+            os.killpg(self.runner_process.pid, signal.SIGKILL)
+
+
+class QEMUTargetRunner(TargetRunner):
+    """
+    Class for interacting with QEMU runners.
+
+    :class:`QEMUTargetRunner` is a subclass of :class:`TargetRunner` which performs necessary
+    groundwork for launching a guest OS on QEMU.
+
+    :param qemu_settings: A dictionary which has QEMU related parameters. The full list
+        of QEMU parameters is below:
+        * ``kernel_image``: This is the location of kernel image (e.g., ``Image``) which
+            will be used as target's kernel.
+
+        * ``arch``: Architecture type. Defaults to ``aarch64``.
+
+        * ``cpu_types``: List of CPU ids for QEMU. The list only contains ``cortex-a72`` by
+            default. This parameter is valid for Arm architectures only.
+
+        * ``initrd_image``: This points to the location of initrd image (e.g.,
+            ``rootfs.cpio.xz``) which will be used as target's root filesystem if kernel
+            does not include one already.
+
+        * ``mem_size``: Size of guest memory in MiB.
+
+        * ``num_cores``: Number of CPU cores. Guest will have ``2`` cores by default.
+
+        * ``num_threads``: Number of CPU threads. Set to ``2`` by defaults.
+
+        * ``cmdline``: Kernel command line parameter. It only specifies console device in
+            default (i.e., ``console=ttyAMA0``) which is valid for Arm architectures.
+            May be changed to ``ttyS0`` for x86 platforms.
+
+        * ``enable_kvm``: Specifies if KVM will be used as accelerator in QEMU or not.
+            Enabled by default if host architecture matches with target's for improving
+            QEMU performance.
+    :type qemu_settings: Dict
+
+    :param connection_settings: the dictionary to store connection settings
+        of :attr:`Target.connection_settings`, defaults to None.
+    :type connection_settings: Dict, optional
+
+    :param make_target: Lambda function for creating :class:`Target` based
+        object, defaults to :func:`lambda **kwargs: LinuxTarget(**kwargs)`.
+    :type make_target: func, optional
+
+    :Variable positional arguments: Forwarded to :class:`TargetRunner`.
+
+    :raises FileNotFoundError: if QEMU executable, kernel or initrd image cannot be found.
+    """
+
+    def __init__(self,
+                 qemu_settings,
+                 connection_settings=None,
+                 # pylint: disable=unnecessary-lambda
+                 make_target=lambda **kwargs: LinuxTarget(**kwargs),
+                 **args):
+        self.connection_settings = {
+            'host': '127.0.0.1',
+            'port': 8022,
+            'username': 'root',
+            'password': 'root',
+            'strict_host_check': False,
+        }
+
+        if connection_settings is not None:
+            self.connection_settings = self.connection_settings | connection_settings
```

Comment:
Union of dict is brought by [PEP 584](https://peps.python.org/pep-0584/), which was implemented in Python 3.8. For now, devlib supports down to 3.7. The backward-compatible equivalent would be:
```
        if connection_settings is not None:
            self.connection_settings = {**self.connection_settings, **connection_settings}
``` 
and you can further avoid a separate path with:
```
self.connection_settings = {**self.connection_settings, **(connection_settings or {})}
```

Reply:
Done.